pkg/eventpersister/persister/persister.go

```diff
@@ -614,3 +548,141 @@ func (p *Persister) key(
 		environmentNamespace,
 	)
 }
+
+// validateTimestamp limits date range of the given timestamp
+func (p *Persister) validateTimestamp(
+	timestamp int64,
+	oldestTimestampDuration, furthestTimestampDuration time.Duration,
+) bool {
+	given := time.Unix(timestamp, 0)
+	maxPast := time.Now().Add(-oldestTimestampDuration)
+	if given.Before(maxPast) {
+		return false
+	}
+	maxFuture := time.Now().Add(furthestTimestampDuration)
+	return !given.After(maxFuture)
+}
+
+func (p *Persister) getEvaluation(
+	ctx context.Context,
+	event *eventproto.GoalEvent,
+	environmentNamespace string,
+) (*featureproto.Evaluation, bool, error) {
+	// List experiments with the following status RUNNING, FORCE_STOPPED, and STOPPED
+	experiments, err := p.listExperiments(ctx, environmentNamespace)
+	if err != nil {
+		return nil, true, err
+	}
+	if len(experiments) == 0 {
+		handledCounter.WithLabelValues(codeNoExperiments).Inc()
+		return nil, false, ErrNoExperiments
+	}
+	// Find the experiment by goal ID
+	// TODO: we must change the console UI not to allow creating
+	// multiple experiments running at the same time,
+	// using the same feature flag id and goal id
+	var experiment *exproto.Experiment
+	for _, exp := range experiments {
+		if p.findGoalID(event.GoalId, exp.GoalIds) {
+			experiment = exp
+			break
+		}
+	}
+	if experiment == nil {
+		handledCounter.WithLabelValues(codeExperimentNotFound).Inc()
+		return nil, false, ErrExperimentNotFound
+	}
+	// For requests with no tag, it will insert "none" instead, until all old SDK clients are updated
+	var tag string
+	if event.Tag == "" {
+		tag = "none"
+	} else {
+		tag = event.Tag
+	}
+	// Get the user evaluation using the experiment info
+	ev, err := p.getUserEvaluation(
+		environmentNamespace,
+		event.UserId,
+		tag,
+		experiment.FeatureId,
+		experiment.FeatureVersion,
+	)
+	if err != nil {
+		return nil, true, err
+	}
+	return ev, false, nil
+}
+
+func (p *Persister) listExperiments(
+	ctx context.Context,
+	environmentNamespace string,
+) ([]*exproto.Experiment, error) {
+	exp, err, _ := p.flightgroup.Do(
+		fmt.Sprintf("%s:%s", environmentNamespace, "listExperiments"),
+		func() (interface{}, error) {
+			experiments := []*exproto.Experiment{}
+			cursor := ""
+			for {
+				resp, err := p.experimentClient.ListExperiments(ctx, &exproto.ListExperimentsRequest{
+					PageSize:             listRequestSize,
+					Cursor:               cursor,
+					EnvironmentNamespace: environmentNamespace,
+					Statuses: []exproto.Experiment_Status{
+						exproto.Experiment_RUNNING,
+						exproto.Experiment_FORCE_STOPPED,
+						exproto.Experiment_STOPPED,
```

Comment:
Because we can still receive delayed goal events, we should count them up to 24hrs after the experiment is finished.

Reply:
listExperiments() should discard the experiments finished more than 24 hours ago?

Reply:
Before we list the experiments, we check the goal event timestamp. If it is more than 24 hours, the event will be acknowledged without being sent to Kafka, so it is ok as it is.
Also, I just noticed that we could list only experiments that are not archived to speed up the linking process. I will fix it. Thanks!